### PR TITLE
Coroutines support for TVM

### PR DESCRIPTION
**Pull request by azhogin**
toolchain-test-suite fix:
https://github.com/tonlabs/toolchain-test-suite/pull/65

### CODE REVIEW

**Comment by akiramenai** [comment]
It seems that includes needs to be updated in the following tests

********************
Failing Tests (3):
    Clang :: CodeGen/tvm/Piggybank.cpp
    Clang :: CodeGen/tvm/Wallet.cpp
    Clang :: CodeGen/tvm/contract-handle-test.cpp

  Expected Passes    : 20
  Unexpected Failures: 3

**Comment by azhogin** [comment]
> It seems that includes needs to be updated in the following tests
> 
> Failing Tests (3):
> Clang :: CodeGen/tvm/Piggybank.cpp
> Clang :: CodeGen/tvm/Wallet.cpp
> Clang :: CodeGen/tvm/contract-handle-test.cpp
> 
> Expected Passes : 20
> Unexpected Failures: 3

Maybe, build system doesn't update branch when I use force-push?

**Comment by akiramenai** [comment]
I rechecked it locally on 
commit 6ac6dd3820b9ad7b776bb55783805dea7056fadb (HEAD -> azhogin/coroutines-support, origin/azhogin/coroutines-support)
It seems that some files still refer to message_flags.hpp
```
llvm/projects/ton-compiler/cpp-sdk/tvm/contract.hpp:#include <tvm/message_flags.hpp>
llvm/projects/ton-compiler/cpp-sdk/tvm/contract_handle.hpp:#include <tvm/message_flags.hpp>
llvm/projects/ton-compiler/cpp-sdk/tvm/smart_switcher.hpp:#include <tvm/message_flags.hpp>
```
But there is no such file anymore

File: llvm/lib/Transforms/Coroutines/CoroElide.cpp [hunk 1]
```diff
@@ -168,10 +168,10 @@ bool Lowerer::shouldElide(Function *F, DominatorTree &DT) const {
   SmallPtrSet<CoroSubFnInst *, 4> DAs;
   for (CoroSubFnInst *DA : DestroyAddr) {
     for (Instruction *TI : Terminators) {
-      if (DT.dominates(DA, TI)) {
+      //if (DT.dominates(DA, TI)) {
```
Comment (akiramenai) [inline llvm/lib/Transforms/Coroutines/CoroElide.cpp:171]: Can we remove commented lines?

Comment (azhogin) [reply llvm/lib/Transforms/Coroutines/CoroElide.cpp]: Fixed.

File: llvm/projects/ton-compiler/cpp-sdk/tvm/contract_handle.hpp [hunk 2]
```diff
@@ -2,8 +2,13 @@
 
 #include <tvm/reflection.hpp>
 #include <tvm/contract.hpp>
+#include <tvm/schema/estimate_element.hpp>
+#include <tvm/chain_tuple.hpp>>
```
Comment (akiramenai) [inline llvm/projects/ton-compiler/cpp-sdk/tvm/contract_handle.hpp:6]: Does it compile?

File: llvm/tools/clang/lib/Driver/ToolChains/TVM.cpp [hunk 3]
```diff
@@ -81,7 +81,7 @@ const char *tvm::Linker::constructClangCommand(
     const llvm::opt::ArgList &Args, llvm::StringRef OutputFilePrefix,
     const char *InputFileName) const {
   // Construct clang command.
-  ArgStringList ClangArgs{InputFileName, "-export-json-abi", "-o"};
+  ArgStringList ClangArgs{InputFileName, "-target", "tvm", "-export-json-abi", "-o"};
```
Comment (akiramenai) [inline llvm/tools/clang/lib/Driver/ToolChains/TVM.cpp:84]: `tvm-unknown-unknown` might help us to suppress
`warning: overriding the module target triple with tvm [-Woverride-module]`

File: llvm/projects/ton-compiler/cpp-sdk/tvm/contract_handle.hpp (hunk 2 above)
Comment (azhogin) [reply llvm/projects/ton-compiler/cpp-sdk/tvm/contract_handle.hpp]: Yes, it worked. Fixed.

File: llvm/tools/clang/lib/Driver/ToolChains/TVM.cpp (hunk 3 above)
Comment (azhogin) [reply llvm/tools/clang/lib/Driver/ToolChains/TVM.cpp]: Removed this change.

**Pull request merged**